tapo/src/requests/set_device_info/trv.rs

```diff
@@ -0,0 +1,75 @@
+use serde::Serialize;
+
+use crate::error::Error;
+
+#[derive(Debug, Default, Serialize)]
+pub(crate) struct TrvSetDeviceInfoParams {
+    #[serde(skip_serializing_if = "Option::is_none", rename = "target_temp")]
+    target_temperature: Option<u8>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    frost_protection_on: Option<bool>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    child_protection: Option<bool>,
+    #[serde(skip_serializing_if = "Option::is_none", rename = "temp_offset")]
+    temperature_offset: Option<i8>,
+    #[serde(skip_serializing_if = "Option::is_none", rename = "min_temp")]
+    min_temperature: Option<u8>,
+    #[serde(skip_serializing_if = "Option::is_none", rename = "min_control_temp")]
+    min_control_temperature: Option<u8>,
+    #[serde(skip_serializing_if = "Option::is_none", rename = "max_control_temp")]
+    max_control_temperature: Option<u8>,
+}
+
+impl TrvSetDeviceInfoParams {
+    pub fn target_temperature(mut self, value: u8) -> Result<Self, Error> {
+        self.target_temperature = Some(value);
+        self.validate()
+    }
+    pub fn frost_protection_on(mut self, value: bool) -> Result<Self, Error> {
+        self.frost_protection_on = Some(value);
+        self.validate()
+    }
+    pub fn child_protection(mut self, value: bool) -> Result<Self, Error> {
+        self.child_protection = Some(value);
+        self.validate()
+    }
+    pub fn temperature_offset(mut self, value: i8) -> Result<Self, Error> {
+        self.temperature_offset = Some(value);
+        self.validate()
+    }
+    pub fn min_control_temperature(mut self, value: u8) -> Result<Self, Error> {
+        self.min_control_temperature = Some(value);
+        self.validate()
+    }
+    pub fn max_control_temperature(mut self, value: u8) -> Result<Self, Error> {
+        self.max_control_temperature = Some(value);
+        self.validate()
+    }
+}
+
+impl TrvSetDeviceInfoParams {
+    pub(crate) fn new() -> Self {
+        Self {
+            target_temperature: None,
+            frost_protection_on: None,
+            child_protection: None,
+            temperature_offset: None,
+            min_temperature: None,
+            min_control_temperature: None,
+            max_control_temperature: None,
+        }
+    }
+    
+    pub fn validate(self) -> Result<Self, Error> {
+        if let Some(temp_offset) = self.temperature_offset {
+            if temp_offset < -10 || temp_offset> 10 {
+                return Err(Error::Validation {
+                    field: "temp_offset".to_string(),
+                    message: "must be between -10 and 10".to_string(),
+                    
+                });
+            }
```

Comment:
The comment of the `set_temperature_offset` says that the temperature offset must be between 5 and 30. But this rule checks that it's between -10 and 10. Which one is accurate?

Could this differ based on the temperature unit?

Reply:
Good catch. The comment is a copy paste issue. I'll fix that. The -10/10 I got from what the app let's me select for the temperature offset. It's fixed to Celsius in my app, but I'd assume that the Fahrenheit values may differ. For some reason I can't get my app to let me select Fahrenheit as the temperature unit.

Reply:
The temperature settings for my other devices are located in the Device Settings.
![image](https://github.com/mihai-dinculescu/tapo/assets/9944681/e1701c97-da9d-481b-aa8b-4627695bbaa8)

Reply:
I don’t get this option for the KE100 device. Very strange. Checked with the android and iOS app and the documentation and there seems to be no option to change that parameter. The temperature unit is not mentioned for this device.

Reply:
Looking at [this video](https://www.youtube.com/watch?v=xStn7GV0PSk), it looks like the unit is shown, but maybe not configurable?

![image](https://github.com/mihai-dinculescu/tapo/assets/9944681/cfd340d0-bd15-4059-958a-33c6d01511ae)
![image](https://github.com/mihai-dinculescu/tapo/assets/9944681/7b27e43b-d19d-4de9-b502-cc75885bca2b)

Reply:
Can you please add `temp_unit` to `TrvSetDeviceInfoParams` and check if setting it does anything?

Reply:
Setting `temp_unit` to "celsius" works. Setting it to "fahrenheit" results in a Error: Tapo(Unknown(-1001)).

```
DEBUG tapo::api::api_client                     > Control child...
DEBUG tapo::api::protocol::passthrough_protocol > Request to passthrough: {"method":"control_child","params":{"device_id":"[REDACTED]","requestData":{"method":"multipleRequest","params":{"requests":[{"method":"set_device_info","params":{"temp_unit":"celsius"}}]}}}}
DEBUG tapo::api::protocol::passthrough_protocol > Device responded with: TapoResponse { error_code: 0, result: Some(TapoResult { response: "[REDACTED]" }) }
DEBUG tapo::api::protocol::passthrough_protocol > Device inner response decrypted: {"result":{"responseData":{"result":{"responses":[{"method":"set_device_info","error_code":0}]}}},"error_code":0}
DEBUG tapo::api::protocol::passthrough_protocol > Device inner response: TapoResponse { error_code: 0, result: Some(ControlChildResult { response_data: TapoMultipleResponse { result: TapoMultipleResult { responses: [TapoResponse { error_code: 0, result: None }] } } }) }

DEBUG tapo::api::api_client                     > Control child...
DEBUG tapo::api::protocol::passthrough_protocol > Request to passthrough: {"method":"control_child","params":{"device_id":"[REDACTED]","requestData":{"method":"multipleRequest","params":{"requests":[{"method":"set_device_info","params":{"temp_unit":"fahrenheit"}}]}}}}
DEBUG tapo::api::protocol::passthrough_protocol > Device responded with: TapoResponse { error_code: 0, result: Some(TapoResult { response: "[REDACTED]" }) }
DEBUG tapo::api::protocol::passthrough_protocol > Device inner response decrypted: {"result":{"responseData":{"result":{"responses":[{"method":"set_device_info","error_code":-1001}]}}},"error_code":0}
DEBUG tapo::api::protocol::passthrough_protocol > Device inner response: TapoResponse { error_code: 0, result: Some(ControlChildResult { response_data: TapoMultipleResponse { result: TapoMultipleResult { responses: [TapoResponse { error_code: -1001, result: None }] } } }) }
Error: Tapo(Unknown(-1001))
```

So my assumption is that it shows the unit in the app and has the field in the API, but `temp_unit` is not configurable for the KE100. I suggest I'll update the documentation in the code regarding this. Maybe worthwhile to also put that in the main readme.

Reply:
Thank you for taking the time to look into this and getting to the bottom.
Thermostat temperature is an important topic, and I agree that the library must be as obvious as possible.

How about adding a `temperature_unit` parameter to all the functions that change temperature (`set_target_temperature`, `set_min_control_temperature` and `set_max_control_temperature.`). It can be a separate enum (like it used to be) with just one value, Celsius.

Forcing users to specify the unit, even if there's only one option, will make it very obvious.

Reply:
Sounds reasonable. I'll update the respective functions.

Reply:
a9ebad1040a2861c931b93b6e11a153d28c49c75 now implements a specific `temperature_unit` parameter for the functions setting a temperature.